lib/internal/modules/esm/loader.js

```diff
@@ -148,29 +147,22 @@ function nextHookFactory(chain, meta, { validateArgs, validateOutput }) {
   }
 
   return ObjectDefineProperty(
-    async (...args) => {
+    async (arg0 = undefined, context) => {
       // Update only when hook is invoked to avoid fingering the wrong filePath
       meta.hookErrIdentifier = `${hookFilePath} '${hookName}'`;
 
-      validateArgs(`${meta.hookErrIdentifier} hook's ${nextHookName}()`, args);
+      validateArgs(`${meta.hookErrIdentifier} hook's ${nextHookName}()`, arg0, context);
 
       const outputErrIdentifier = `${chain[generatedHookIndex].url} '${hookName}' hook's ${nextHookName}()`;
 
       // Set when next<HookName> is actually called, not just generated.
       if (generatedHookIndex === 0) { meta.chainFinished = true; }
 
-      // `context` is an optional argument that only needs to be passed when changed
-      switch (args.length) {
-        case 1: // It was omitted, so supply the cached value
-          ArrayPrototypePush(args, meta.context);
-          break;
-        case 2: // Overrides were supplied, so update cached value
-          ObjectAssign(meta.context, args[1]);
-          break;
+      if (context) { // `context` has already been validated, so no fancy check needed.
+        ObjectAssign(meta.context, context);
       }
 
-      ArrayPrototypePush(args, nextNextHook);
-      const output = await ReflectApply(hook, undefined, args);
+      const output = await hook(arg0, context, nextNextHook);
```

Comment:
To fix the failing test, when `context` is nullish, we should pass `meta.context` to the next hook.
```suggestion
      const output = await hook(arg0, context ?? meta.context, nextNextHook);
```

Reply:
Probably a topic for another ticket, but if a misbehaving hook passes `false` for `context`, this code is going to behave in a sorta unexpected way.  I'm not totally sure the intent here, fields are copied off of `context` onto `meta.context`, but `context` value is passed to the next hook instead of `meta.context`?  We check truthyness in the `if()` above, and then check nullishness here?

Reply:
OH! whoops. this should be `meta.context` regardless

Reply:
@cspotcode when `hook()` is passed `meta.context`, there's no issue when `context` is `false` (it's just ignored).

Reply:
> Probably a topic for another ticket, but if a misbehaving hook passes `false` for `context`, this code is going to behave in a sorta unexpected way. I'm not totally sure the intent here, fields are copied off of `context` onto `meta.context`, but `context` value is passed to the next hook instead of `meta.context`? We check truthyness in the `if()` above, and then check nullishness here?

I agree, `validateArgs` should throw if `context` is neither an object or nullish, but it looks like it accepts any falsy value.

Reply:
> OH! whoops. this should be `meta.context` regardless

Note that the previous code was more or less doing `context ?? meta.context`. I'm not sure what difference does this make, I suppose if `context` inherits some values from a prototype that could have an effect. I'm not against passing `meta.context` regardless, but wanted to point out in case the previous implementation was done on purpose.

Reply:
Mm, I mentioned the difference in an earlier comment somewhere else in this PR 🙂

I'm pretty sure the end-result is the same; this way is just less code and more straightforward.

In a separate PR I can update the validateArgs fns to specifically check for not undefined or not nullish (so that `false` triggers a validation failure).

Reply:
Btw @aduh95 could you please re-approve 🙂